x/rollapp/transfergenesis/ibc_module.go

```diff
@@ -108,43 +119,122 @@ func (w IBCModule) OnRecvPacket(
+	// handle cases where genesis transfer is not expected (post genesis / no IRO plan)
+	if !w.IsGenesisTransferExpected(ctx, ra) {
+		// genesis transfer not expected but present. handle DRS violation
+		if isGenesisTransfer {
+			l.Error("Genesis transfer not expected.")
+			_ = w.handleDRSViolation(ctx, ra.RollappId)
+			return uevent.NewErrorAcknowledgement(ctx, ErrDisabled)
 		}
-		return uevent.NewErrorAcknowledgement(ctx, ErrDisabled)
+
+		// first transfer when genesis transfer is not required should enable transfers
+		if !ra.IsGenesisTransferEnabled() {
+			err := w.EnableTransfers(ctx, ra.RollappId, transfer.Denom)
+			if err != nil {
+				l.Error("Enable transfers.", "err", err)
+				return uevent.NewErrorAcknowledgement(ctx, errorsmod.Wrap(err, "transfer genesis: enable transfers"))
+			}
+		}
+
+		// post genesis case - continue with the normal IBC flow
+		return w.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	/* ------------------------ genesis transfer required ----------------------- */
+	if !isGenesisTransfer {
+		l.Error("genesis transfer required.")
+		return uevent.NewErrorAcknowledgement(ctx, errorsmod.Wrap(gerrc.ErrFailedPrecondition, "genesis transfer required"))
+	}
+
+	// handle genesis transfer by the IRO keeper
+	plan := w.iroKeeper.MustGetPlanByRollapp(ctx, ra.RollappId)
+
+	// validate the transfer against the IRO plan
+	err = w.validateGenesisTransfer(plan, transfer, memo.Denom)
+	if err != nil {
+		l.Error("Validate IRO plan.", "err", err)
+		return uevent.NewErrorAcknowledgement(ctx, errorsmod.Wrap(err, "validate IRO plan"))
 	}
-	// it's a valid genesis transfer!
 
 	err = w.registerDenomMetadata(ctx, ra.RollappId, ra.ChannelId, memo.Denom)
 	if err != nil {
 		l.Error("Register denom metadata.", "err", err)
 		return uevent.NewErrorAcknowledgement(ctx, errorsmod.Wrap(err, "transfer genesis: register denom metadata"))
 	}
 
-	l.Debug("Received valid genesis transfer. Registered denom data.")
+	// set the ctx to skip delayedack etc because we want the transfer to happen immediately
+	ack = w.IBCModule.OnRecvPacket(commontypes.SkipRollappMiddlewareContext(ctx), packet, relayer)
+	// if the ack is nil, we return an error as we expect immediate ack
+	if ack == nil {
+		l.Error("Expected immediate ack for genesis transfer.")
+		return uevent.NewErrorAcknowledgement(ctx, errorsmod.Wrap(gerrc.ErrInternal, "transfer genesis: OnRecvPacket"))
+	}
+	err = w.EnableTransfers(ctx, ra.RollappId, transfer.Denom)
+	if err != nil {
+		l.Error("Enable transfers.", "err", err)
+		return uevent.NewErrorAcknowledgement(ctx, errorsmod.Wrap(err, "transfer genesis: enable transfers"))
+	}
+	return ack
+}
+
+// IsGenesisTransferExpected checks if the genesis transfer is expected for the rollapp
+// if the prelaunch time is set, then genesis transfer is expected as we have IRO plan in place
+func (w IBCModule) IsGenesisTransferExpected(ctx sdk.Context, rollapp *rollapptypes.Rollapp) bool {
+	if rollapp.GenesisState.TransfersEnabled {
```

Comment:
why dont we check in the beginning of the OnRecvPacket right after we know it's a rollapp (and if true, contine to next) vs checking if we expect a transfer.

Reply:
so we could detect violation where `TransfersEnabled=true` and we have genesis transfer.
as we discussed it can be removed, I'll refactor